src/instamatic/experiments/fast_adt/experiment.py

```diff
@@ -0,0 +1,452 @@
+from __future__ import annotations
+
+import logging
+from dataclasses import dataclass, field
+from pathlib import Path
+from queue import Queue
+from threading import Thread
+from typing import Any, Dict, Iterator, List, Optional, Sequence, Tuple, Union
+
+import numpy as np
+import pandas as pd
+from matplotlib import pyplot as plt
+from typing_extensions import Self
+
+from instamatic import config
+from instamatic._collections import NullLogger
+from instamatic._typing import AnyPath
+from instamatic.calibrate import CalibBeamShift, CalibMovieDelays, CalibStageRotation
+from instamatic.calibrate.filenames import CALIB_BEAMSHIFT
+from instamatic.experiments.experiment_base import ExperimentBase
+from instamatic.processing.ImgConversionTPX import ImgConversionTPX as ImgConversion
+
+
+def safe_range(start: float, stop: float, step: float) -> np.ndarray:
+    """Find 2+ floats between `start` and `stop` (inclusive) ~`step` apart."""
+    step_count = max(round(abs(stop - start) / step) + 1, 2)
+    return np.linspace(start, stop, step_count, endpoint=True, dtype=float)
+
+
+class FastADTEarlyTermination(RuntimeError):
+    """Raised if FastADT experiment terminates early due to known reasons."""
+
+
+class FastADTMissingCalibError(RuntimeError):
+    """Raised if some calibration is strictly required but missing."""
```

Comment:
Imo these scream to be generalized (perhaps at a later stage).

```suggestion
class EarlyTermination(RuntimeError):
    """Raised if experiment terminates early due to known reasons."""


class MissingCalibError(RuntimeError):
    """Raised if some calibration is strictly required but missing."""
```